Add GPU health check

guptaNswati opened the pull request:
Addressing https://github.com/NVIDIA/k8s-dra-driver-gpu/issues/360 to add preliminary health check similar to https://github.com/NVIDIA/k8s-device-plugin. 

- Clean follow-up of https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/545 
- addressing review comments 
- republising of resourceslice on health event 
- feature gate: --set featureGates.DeviceHealthCheck=true
- xids: --set kubeletPlugin.gpus.additionalXidsToIgnore="n1,n2"

Test logs:
```
I1017 20:16:19.799738       1 device_health.go:179] Processing event {Device:{Handle:0xe4aea6b2fef0} EventType:8 EventData:43 GpuInstanceId:7 ComputeInstanceId:0}
I1017 20:16:19.799821       1 device_health.go:192] Event for mig device: &{<nil> 0x40006a0070 Healthy}
I1017 20:16:19.799843       1 device_health.go:202] Sending unhealthy notification for device MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 due to event type: 8 and event data: 43
W1017 20:16:19.799870       1 driver.go:219] Received unhealthy notification for device: MIG-4d806f22-346a-5a1d-ac01-86b505cdf485
I1017 20:16:19.799884       1 device_state.go:558] Update device sattus:MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 healthstatus
I1017 20:16:19.799891       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x400040c150 Healthy}
I1017 20:16:19.799955       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a0ee0 Healthy}
I1017 20:16:19.799966       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a0f50 Healthy}
I1017 20:16:19.799974       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a0230 Healthy}
I1017 20:16:19.799983       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a02a0 Healthy}
I1017 20:16:19.799992       1 driver.go:230] Device is healthy, added to resoureslice: &{0x40002ae000 <nil> Healthy}
I1017 20:16:19.800000       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x400040c0e0 Healthy}
W1017 20:16:19.800009       1 driver.go:233] Device:MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 with uuid:&{%!s(*main.GpuInfo=<nil>) %!s(*main.MigDeviceInfo=&{MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 1g.12gb 0x40002503c0 0x400049c130 0x4000610090 0x400045ce40 0x4000610150 0x40004840f0 0009:01:00.0 0x40002fbf50}) Unhealthy} is unhealthy
I1017 20:16:19.800021       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a00e0 Healthy}
I1017 20:16:19.800031       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x400040c1c0 Healthy}
I1017 20:16:19.800043       1 driver.go:230] Device is healthy, added to resoureslice: &{0x40002503c0 <nil> Healthy}
I1017 20:16:19.800050       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x400040c000 Healthy}
I1017 20:16:19.800056       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a0310 Healthy}
I1017 20:16:19.800063       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a0150 Healthy}
I1017 20:16:19.800069       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x400040c070 Healthy}
I1017 20:16:19.800076       1 driver.go:230] Device is healthy, added to resoureslice: &{<nil> 0x40006a01c0 Healthy}
I1017 20:16:19.800084       1 driver.go:237] [Rebulishing resourceslice with healthy devices
I1017 20:16:19.800142       1 driver.go:247] Successfully republished resources without unhealthy device MIG-4d806f22-346a-5a1d-ac01-86b505cdf485:
I1017 20:16:19.800178       1 resourceslicecontroller.go:647] "Existing slices" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" obsolete=[] current=["sc-starwars-mab9-b00-gpu.nvidia.com-kmcts"]
I1017 20:16:19.800209       1 resourceslicecontroller.go:724] "Need to update slice" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" slice="sc-starwars-mab9-b00-gpu.nvidia.com-kmcts" matchIndex=0
I1017 20:16:19.800225       1 resourceslicecontroller.go:727] "Completed comparison" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" numObsolete=0 numMatchedSlices=1 numChangedMatchedSlices=1 numNewSlices=0
I1017 20:16:19.800230       1 resourceslicecontroller.go:743] "Kept generation because at most one update API call is necessary" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" generation=1
I1017 20:16:19.805795       1 round_trippers.go:632] "Response" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" verb="PUT" url="https://10.96.0.1:443/apis/resource.k8s.io/v1beta1/resourceslices/sc-starwars-mab9-b00-gpu.nvidia.com-kmcts" status="200 OK" milliseconds=5
I1017 20:16:19.806290       1 resourceslicecontroller.go:779] "Updated existing resource slice" logger="ResourceSlice controller" poolName="sc-starwars-mab9-b00" slice="sc-starwars-mab9-b00-gpu.nvidia.com-kmcts"
I1017 20:16:19.807922       1 resourceslicecontroller.go:500] "ResourceSlice update" logger="ResourceSlice controller" slice="sc-starwars-mab9-b00-gpu.nvidia.com-kmcts" diff=<
	@@ -3,8 +3,8 @@
	   "name": "sc-starwars-mab9-b00-gpu.nvidia.com-kmcts",
	   "generateName": "sc-starwars-mab9-b00-gpu.nvidia.com-",
	   "uid": "7184f664-55c1-412d-bd99-4b46e7c23846",
	-  "resourceVersion": "59000652",
	-  "generation": 1,
	+  "resourceVersion": "59001011",
	+  "generation": 2,
	   "creationTimestamp": "2025-10-17T20:14:42Z",
	   "ownerReferences": [
	    {
	@@ -20,7 +20,7 @@
	     "manager": "gpu-kubelet-plugin",
	     "operation": "Update",
	     "apiVersion": "resource.k8s.io/v1beta1",
	-    "time": "2025-10-17T20:14:42Z",
	+    "time": "2025-10-17T20:16:19Z
	.....
	.....
	
		     "name": "gpu-1",
	     "attributes": {
	      "architecture": {
	@@ -161,7 +496,7 @@
	     }
	    },
	    {
	-    "name": "gpu-0-mig-19-0-1",
	+    "name": "gpu-0-mig-19-1-1",
	     "attributes": {
	      "architecture": {
	       "string": "Hopper"
	@@ -197,56 +532,230 @@
	       "string": "mig"
	      },
	      "uuid": {
	-      "string": "MIG-4d806f22-346a-5a1d-ac01-86b505cdf485"
	-     }
	-    },
```
TLDR, This device had an event and is not added `MIG-4d806f22-346a-5a1d-ac01-86b505cdf485`

The device is picked back when driver is restarted.

shivamerla commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,281 @@
+			var affectedDevice *AllocatableDevice
+			if event.GpuInstanceId != FullGPUInstanceID && event.ComputeInstanceId != FullGPUInstanceID {
+				affectedDevice = m.findMigDevice(eventUUID, event.GpuInstanceId, event.ComputeInstanceId)
+				klog.Infof("Event for mig device: %s", affectedDevice.UUID())
```
```suggestion
                if affectedDevice != nil {
				    klog.Infof("Event for mig device: %s", affectedDevice.UUID())
				}
```

shivamerla commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,281 @@
+				klog.Infof("Event for mig device: %s", affectedDevice.UUID())
+			} else {
+				affectedDevice = m.findGpuDevice(eventUUID)
+				klog.Infof("Event for device: %s", affectedDevice.UUID())
```
```suggestion
                if affectedDevice != nil {
				    klog.Infof("Event for device: %s", affectedDevice.UUID())
				}
```

shivamerla commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,281 @@
+				continue
+			}
+
+			if event.EventType != nvml.EventTypeXidCriticalError {
```
We are setting event mask on XID critical errors as well as SBE and DBE errors when we subscribe for events. Similarly should we also handle those here instead of skipping those by default?

elezar replied:
Yes, I have noted this in the device plugin too. I would assume that the device plugin implementation also needs to be updated, but since I've ever seen these events in the "wild", assessing the current behaviour is non-trivial.

shivamerla commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,54 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+				}
+			}
+
+			klog.V(6).Info("Rebulishing resourceslice with healthy devices")
```
if all devices are unhealthy, we will be publishing a resource slice with empty devices. Is that intended? Also, should we synchronize this with a concurrent node prepare/unprepare calls happening (i.e prep/unprepare locks)?

shivamerla replied:
Thanks for clarifying. I suppose the advantage of holding the prep/unprep locks while updating the global device state is that any pending Prepare call will correctly fail the allocation if the device becomes unhealthy, instead of assigning an unhealthy GPU to the workload. This way, the claim remains in the pending state, allowing users to reschedule on other nodes (e.g., by re-creating the claims or through templates).

shivamerla commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,281 @@
+			if ret != nvml.SUCCESS {
+				klog.Infof("Error waiting for event: %v; Marking all devices as unhealthy", ret)
+				for _, dev := range m.uuidToDeviceMap {
+					m.unhealthy <- dev
```
Should we use a non blocking send here if the event handler is busy and drop those events? This depends on if the events are repetitive so we get to handle those eventually. This might also avoid any potential deadlock scenarios.

guptaNswati commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,281 @@
+			klog.Infof("Processing event %+v", event)
+			eventUUID, ret := event.Device.GetUUID()
+			if ret != nvml.SUCCESS {
+				klog.Infof("Failed to determine uuid for event %v: %v; Marking all devices as unhealthy.", event, ret)
```
its how its done in device-plugin https://github.com/NVIDIA/k8s-device-plugin/blob/main/internal/rm/health.go#L147

elezar commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+				klog.V(6).Info("Successfully republished resources without unhealthy device")
+			}
+
+			release()
```
If we factor this logic into a function, then we could use a deferred call to `release()` after taking the lock. This may be less error prone if we do ever add code paths that return from this logic.

elezar commented on cmd/gpu-kubelet-plugin/main.go:
```diff
@@ -147,6 +148,13 @@ func newApp() *cli.App {
+			Value:       "",
+			Destination: &flags.additionalXidsToIgnore,
+			EnvVars:     []string{"ADDITIONAL_XIDs_TO_IGNORE"},
+		},
```
In https://github.com/NVIDIA/k8s-device-plugin/pull/1443 we added a list of EXPLICIT XIDs to consider fatal. This allows a user to:
1. Specify ignored XIDs (including `all`)
2. Specify SPECIFIC XIDs that are considered fatal (including `all`).

The important thing here is that it allows users to  override the list of hard-coded XIDs that we currently track.

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,286 @@
+	klog.V(6).Info("registering NVML events for device health monitor")
+	m.registerDevicesForEvents()
+
+	skippedXids := m.xidsToSkip(config.flags.additionalXidsToIgnore)
```
In the device plugin implementation setting skippedIDs to `all` disables health checking. Why is this logic different here?

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,286 @@
+	wg              sync.WaitGroup
+}
+
+func newDeviceHealthMonitor(ctx context.Context, config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (*deviceHealthMonitor, error) {
```
This doesn't just create the health monitor, it also starts it. Why not separate construction from actually starting the monitor? This will allow us to more easily swap out monitors (e.g. with NVSentinel) once we have those ready.

elezar replied:
QUestion: What functionality is required for a `DeviceHealthMonitor` from the perspective of the driver. Would returning the following `interface` here make sense:

```go
type DeviceHealthMonitor interface {
    Start() error
    Stop()
    Unhealthy() <-chan *AllocatableDevice
}
```

jgehrcke replied:
Great discussion! Let's leave this thread open. I think we can get back to this after merging the patch. Vetos?

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,286 @@
+	FullGPUInstanceID uint32 = 0xFFFFFFFF
+)
+
+type deviceHealthMonitor struct {
```
```suggestion
type nvmlDeviceHealthMonitor struct {
```

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,286 @@
+	}
+	m.eventSet = eventSet
+
+	m.uuidToDeviceMap = getUUIDToDeviceMap(allocatable)
```
Question: Does `allocatable` change at all, or is it constant for the lifetime of the plugin?

elezar replied:
Sorry, that wasn't clear. Does the content of `allocatable` change in any way that would invalidate the map that we construct here meaning that it would need to be reconstructed.

guptaNswati replied:
No, the only changing content of an allocatable device is its health status and that wont impact it. This map (`[uuid] = device`) is constructed in the very beginning when the plugin is started. 

and we iterate on it to register event on each device (currently, we dont check any status here. we may do it in future on remediation) and send unhealthy notification.

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,286 @@
+			}
+			if ret != nvml.SUCCESS {
+				klog.Infof("Error waiting for event: %v; Marking all devices as unhealthy", ret)
+				for _, dev := range m.uuidToDeviceMap {
```
As a thought: adding a function `m.markAllDevicesUnhealthy()` may improve readability where we need this logic.

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+			}
+
+			if err := d.pluginhelper.PublishResources(ctx, resources); err != nil {
+				klog.Errorf("Failed to publish resources after device health status update: %v", err)
```
Naturally, I wonder why this error is only handled by logging a message. This might be the correct (or currently best) decision. But please walk the reader of the code through the arguments for ending up with that decision, using a brief code comment.

I'd like to understand thoughts here in the lines of "do not retry, because" or "this is implicitly retried later, because" or "we could crash the plugin here, but" or "the old resource slice state remains published, which is good enough", and so on. I am sure you've thought through all this, and it will pay off to document the argumentation for our future selves.

ArangoGutierrez replied:
Jan-Philip Comment is a request for a code-doc-comment, to save the why retying makes sense for future code maintainers

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+				klog.V(6).Info("Health monitor channel closed")
+				return
+			}
+			uuid := device.UUID()
```
Can `device.UUID()` fail? Edit: I see, it doesn't -- cf `func (d AllocatableDevice) UUID() string`.

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+			klog.Warningf("Received unhealthy notification for device: %s", uuid)
+
+			if !device.IsHealthy() {
+				klog.V(6).Infof("Device: %s is aleady marked unhealthy. Skip republishing resourceslice", uuid)
```
In practice, how often could we see a log message like this?

What I see here right now: we can get the `d.deviceHealthMonitor.Unhealthy()` event multiple times, even if we had already processed that device before. I wonder how often we should expect that to happen.

guptaNswati replied:
there can be events burst. @lalitadithya showed me logs of device-plugin xid errors in a cluster which clearly showed same event logged multiple times. 

@lalitadithya is it possible to share the log here.

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+				continue
+			}
+
+			release, err := d.pulock.Acquire(ctx, flock.WithTimeout(10*time.Second))
```
Oh! Acquiring this lock here is a big decision.

Here, I really expect a concise / precise code comment explaining convincingly

- why this lock must be acquired
- how we guarantee that `release()` is always called

Maybe start by explaining what you think will go wrong when we do _not_ acquire this lock here at this point.

guptaNswati replied:
This lock will prevent unhealthy device to be allocated in a simultaneous NodePrepare call().

guptaNswati replied:
i need to double check that lock is released on any failures.

jgehrcke replied:
We should be mindful of acquiring this lock only for a strong reason. When we introduced that lock we named it prepare/unprepare lock because it's meant for that purpose.

Maybe we should use it here, too -- but let's pretty please thoroughly identify that strong reason, and put it into a few English sentences that are convincing.

I am not yet satisfied yet here by our arguments. We need to discuss the alternatives considered, I need more help please to understand why this is the correct approach (I really mean that -- it's not that I know what we should do -- but I sense that we don't, as a collective, understand yet what we really want to do here).

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+		case <-ctx.Done():
+			klog.V(6).Info("Stop processing device health notifications")
+			return
+		case device, ok := <-d.deviceHealthMonitor.Unhealthy():
```
What do you think about wrapping the logic in this case in a function and then call that function here. I think that would greatly help readability. It may also help with calling `release()` more reliably (if needed -- see discussion [here](https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/689/files#r2454590696)).

guptaNswati commented:
> Thanks for the patience in waiting on a review @guptaNswati.

Thanks to you for the review @elezar. 

Thanks to @jgehrcke also. 

The PR looks more lively :-D

guptaNswati commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+			return
+		case device, ok := <-d.deviceHealthMonitor.Unhealthy():
+			if !ok {
+				klog.V(6).Info("Health monitor channel closed")
```
yes it should only happen once. I remember testing it by killing the driver ds.

guptaNswati commented on cmd/gpu-kubelet-plugin/device_state.go:
```diff
@@ -550,6 +550,14 @@ func GetOpaqueDeviceConfigs(
+	defer s.Unlock()
+
+	device.Health = healthstatus
+	klog.Infof("Update device sattus:%s healthstatus", device.UUID())
```
right now, its only logged when healthy becomes unhealthy. But in future when we have remediation, it will also change to unhealthy->healthy.

guptaNswati commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+			if err := d.pluginhelper.PublishResources(ctx, resources); err != nil {
+				klog.Errorf("Failed to publish resources after device health status update: %v", err)
+			} else {
+				klog.V(6).Info("Successfully republished resources without unhealthy device")
```
i dint purposefully add the UUIDs of unhealthy devices as its already logged multiple times and when ResourceSlice is republished it logs the diff which shows it too.

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,309 @@
+			klog.V(6).Info("Stopping event-driven GPU health monitor...")
+			return
+		default:
+			event, ret := m.eventSet.Wait(5000)
```
I needed to look this up to understand what it's doing.

Ref docs are here: https://docs.nvidia.com/deploy/nvml-api/group__nvmlEvents.html#group__nvmlEvents_1g9714b0ca9a34c7a7780f87fee16b205c

argument is timeout in ms.

errors to be handled:

<img width="901" height="228" alt="Image" src="https://github.com/user-attachments/assets/c218a652-d9d9-4f3b-8331-875b574cb991" />

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,309 @@
+				continue
+			}
+			if ret != nvml.SUCCESS {
+				klog.Infof("Error waiting for event: %v; Marking all devices as unhealthy", ret)
```
This does not seem right. 

Let us specifically handle the `NVML_ERROR_GPU_IS_LOST` case, and perform this 'nuke option' only then. Then we can also have a more precise log message (emitted on error level).

guptaNswati replied:
while this will allow for proper error handling, this will also deviate from how its done in device-plugin. i can take it up as a follow-up to fix in both. 

@jgehrcke thoughts?

guptaNswati replied:
Pinging all the reviewers to help resolve this. Most other seems to be non-blocking. 

@jgehrcke @elezar @shivamerla @ArangoGutierrez

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,309 @@
+				continue
+			}
+
+			klog.Infof("Processing event %+v", event)
```
Do you have an example for this log message, how it would look like in practice?

elezar commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -39,11 +41,13 @@ import (
+	state                   *DeviceState
+	pulock                  *flock.Flock
+	healthcheck             *healthcheck
+	nvmlDeviceHealthMonitor *nvmlDeviceHealthMonitor
```
Since we know that we will add new health monitors, let's rename the member to be more generic.
```suggestion
	deviceHealthMonitor *nvmlDeviceHealthMonitor
```

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,316 @@
+	go m.run(ctx, skippedXids)
+
+	return m, nil
+}
```
As mentioned at the call site, I think it simplifies the implmentation if we split the construction of a monitor from actually starting it. What about updating this to:

```suggestion
func newNvmlDeviceHealthMonitor(config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (*nvmlDeviceHealthMonitor, error) {
	if nvdevlib.nvmllib == nil {
		return nil, fmt.Errorf("nvml library is nil")
	}

	if ret := nvdevlib.nvmllib.Init(); ret != nvml.SUCCESS {
		return nil, fmt.Errorf("failed to initialize NVML: %v", ret)
	}
	defer func() {
		_ = nvdevlib.nvmllib.Shutdown()
	}()

	m := &nvmlDeviceHealthMonitor{
		nvmllib:                  nvdevlib.nvmllib,
		unhealthy:                make(chan *AllocatableDevice, len(allocatable)),
		uuidToDeviceMap:          getUUIDToDeviceMap(allocatable),
		getDeviceByParentGiCiMap: getDeviceByParentGiCiMap(allocatable),
		skippedXids:              xidsToSkip(config.flags.additionalXidsToIgnore),
	}

	return m, nil
}

func (m *nvmlDeviceHealthMonitor) Start(ctx context.Context) (rerr error) {
	if ret := m.nvmllib.Init(); ret != nvml.SUCCESS {
		return fmt.Errorf("failed to initialize NVML: %v", ret)
	}
    // We shutdown nvml if this function returns with an error.
    defer func() {
        if rerr != nil {
            _ = m.nvmllib.Shutdown()
        }
    }()

	klog.V(6).Info("creating NVML events for device health monitor")
	eventSet, ret := m.nvmllib.EventSetCreate()
	if ret != nvml.SUCCESS {
		return fmt.Errorf("failed to create event set: %w", ret)
	}

	ctx, cancel := context.WithCancel(ctx)

	m.cancelContext = cancel
	m.eventSet = eventSet

	klog.V(6).Info("registering NVML events for device health monitor")
	m.registerEventsForDevices()

	klog.V(6).Info("started device health monitoring")
	m.wg.Add(1)
	go m.run(ctx, m.skippedXids)

	return nil
}
```

Note that we now cleanly separate nvml errors that occur during setup and those that we have to handle while waiting for events.

jgehrcke replied:
Was this done? Should this be resolved? @guptaNswati a tiny "done, resolving" always helps :)

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,316 @@
+		45,  // Preemptive cleanup, due to previous errors
+		68,  // Video processor exception
+		109, // Context Switch Timeout Error
+	}
```
I know that this list is taken from the device plugin, but handling them explicity at such a low-level is quite difficult to customize. Does it make sense to *not* port this logic over by instead define these as the default value for the envvar where we expose this to the user?

(I'm happy to do this as a follow-up though).

Just as a note for completeness. The GKE device plugin doesn't use this strategy for XIDs. They only element in their list is XID48 (see https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/0509b1f9f4b9a357b44ba65e7b508ded8bd5ecf0/pkg/gpu/nvidia/health_check/health_checker.go#L59).

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,316 @@
+	wg                       sync.WaitGroup
+}
+
+func newNvmlDeviceHealthMonitor(ctx context.Context, config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (*nvmlDeviceHealthMonitor, error) {
```
This is not a blocking comment for this PR. I think it's an important follow-up, but can be done once we start adding a new HealthMonitor implementation.

elezar commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -93,6 +97,21 @@ func NewDriver(ctx context.Context, config *Config) (*driver, error) {
+		go func() {
+			defer driver.wg.Done()
+			driver.deviceHealthEvents(ctx, config.flags.nodeName)
+		}()
```
> @elezar changing from struct to interface and adding this abstraction to make the monitor more generic will change the whole implementation. Ofcourse there are pros and cons of both approaches but I dont see any problem with the way its done right now.

I don't agree that it "changes the whole implementation". All we are doing is splitting the construction and starting the monitor. Even this is not strictly required if we work against the following interface:
```type HealthMonitor interface {
   Unhealthy() <-chan *AllocatableDevice
   Stop()
}
```
we could return a health monitor that is already started.

And on "Ofcourse there are pros and cons of both approaches but I dont see any problem with the way its done right now." the motivation for adding the abstraction is that we *know* that we want to implement at least two different health monitors (with a third concrete example also discussed). *Knowing* that we want to add a new thing similar to what we are adding now should drive us to make the changes now so that the changes that we make adding the next one are reduced. For example, assuming that we have implemented:
1. A factory method that returns an interface type (or even a concrete type `*nvmlDeviceHealthMonitor`)
2. Use a small set of functions on this type from the DRA driver.

Then adding a new health monitor will not require change to the DRA driver itself, but these will be limited to the factory method internals and the actual implementation. (See https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

> In the new health API, the DRA driver will not be responsible to start or stop the health monitor. I don't think we need this abstraction.

Are you saying that there will not be seperate setup and cleanup required for ANY health monitors? The NVML-based health monitor is at least one example where this is required. At the very least I would expect the DRA driver to be able to invoke some `Cleanup` or `Stop` function as it is terminating.

guptaNswati commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,316 @@
+
+	m.wg.Wait()
+
+	_ = m.eventSet.Free()
```
updated.

guptaNswati commented on deployments/helm/nvidia-dra-driver-gpu/values.yaml:
```diff
@@ -223,6 +224,7 @@ kubeletPlugin:
       # Port running a gRPC health service checked by a livenessProbe.
       # Set to a negative value to disable the service and the probe.
       healthcheckPort: 51516
+      additionalXidsToIgnore: ""
```
removed from values.yaml

guptaNswati commented:
@elezar @jgehrcke i addressed the critical comments and added to-dos for follow-up. Please review again. 

PS. i was cautious to not close the comments :-D

guptaNswati commented on pkg/featuregates/featuregates.go:
```diff
@@ -36,6 +36,9 @@ const (
 	IMEXDaemonsWithDNSNames featuregate.Feature = "IMEXDaemonsWithDNSNames"
+
+	// DeviceHealthCheck allows Device Health Checking.
+	DeviceHealthCheck featuregate.Feature = "DeviceHealthCheck"
```
i would like to keep this generic and rather have a separate way to specify the monitoring provider.

ArangoGutierrez replied:
I disagree with that, as per the discussion we agreed on having NVML "gated" by default, but what you are implying here is that ALL health providers will be gated by default, and a second flag will be needed to pass the name of the provider.

Here we are gating ONLY NVML, as later on we will have NVSentinel via the API (that is under discussion) as the default. 

Feature gates are to modify a default behaviour or enable/disable a feature that by default doesn't work. That is why I think this featureGate name should be NVML tailored.

elezar commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+				}
+				klog.Warningf("Error waiting for NVML event: %v. Retrying...", ret)
+				continue
+			}
```
```suggestion
			if ret == nvml.ERROR_GPU_IS_LOST {
				klog.Infof("GPU is lost error: %v; Marking all devices as unhealthy", ret)
				m.markAllDevicesUnhealthy()
				continue
			}
			if ret != nvml.SUCCESS {
				klog.Warningf("Error waiting for NVML event: %v. Retrying...", ret)
				continue
			}
```

elezar commented on cmd/gpu-kubelet-plugin/device_state.go:
```diff
@@ -296,6 +296,12 @@ func (s *DeviceState) prepareDevices(ctx context.Context, claim *resourceapi.Res
 		}
+		// only proceed with config mapping if device is healthy.
+		if featuregates.Enabled(featuregates.DeviceHealthCheck) {
+			if device.Health == Unhealthy {
```
Why not:

```suggestion
			if !device.IsHealthy() {
```

guptaNswati replied:
good catch.

elezar commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -93,6 +103,23 @@ func NewDriver(ctx context.Context, config *Config) (*driver, error) {
+			defer driver.wg.Done()
+			driver.deviceHealthEvents(ctx, config.flags.nodeName)
+		}()
+	}
```
Instead of having to check `d.deviceHealthMonitor != nil` throughout the code, how about implementing a `noopHealthMonitor` (or `disabledHealthMonitor`) as follows:

```
type disabledDeviceHealthMonitor struct{}

var _ deviceHealthMonitor = (*disabledDeviceHealthMonitor)(nil)

func (m *disabledDeviceHealthMonitor) Start(_ context.Context) error {
	return nil
}

func (m *disabledDeviceHealthMonitor) Stop() {
}

func (m *disabledDeviceHealthMonitor) Unhealthy() <-chan *AllocatableDevice {
	return nil
}
```

Then we can do the following:
```suggestion
	deviceHealthMonitor, err := newDeviceHealthMonitor(config, state.allocatable, state.nvdevlib)
	if err != nil {
		return nil, fmt.Errorf("failed to create NVML device health monitor: %w", err)
	}
	driver.deviceHealthMonitor = deviceHealthMonitor
	if err := deviceHealthMonitor.Start(ctx); err != nil {
		return nil, fmt.Errorf("failed to start device health monitor: %w", err)
	}
	
	driver.wg.Add(1)
	go func() {
		defer driver.wg.Done()
		driver.deviceHealthEvents(ctx, config.flags.nodeName)
	}()
```

with:
```
func newDeviceHealthMonitor(config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (deviceHealthMonitor, error) {
	if !featuregates.Enabled(featuregates.DeviceHealthCheck) {
		return &disabledDeviceHealthMonitor{}, nil
	}

	return newNvmlDeviceHealthMonitor(config, allocatable, nvdevlib)
}
```

guptaNswati replied:
mmm i know its old school but i find the explicit checks more readable. and it also aligns with how the rest of the dra driver code is implemented.

elezar commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -38,12 +40,20 @@ import (
+type deviceHealthMonitor interface {
+	Start(context.Context) error
+	Stop()
+	Unhealthy() <-chan *AllocatableDevice
```
(I don't think this works with the current implementation though since we can't modify the `AllocatableDevice` in the health checking code).

guptaNswati commented on deployments/helm/nvidia-dra-driver-gpu/values.yaml:
```diff
@@ -67,7 +67,8 @@ resources:
 #   LoggingBetaOptions: true           # Kubernetes logging beta features
-featureGates: {}
+featureGates:
+  DeviceHealthCheck: false
```
I know the IMEX one is default but the MPS and timeslicing ones need to enabled on some versions. @klueska can answer this better.

guptaNswati replied:
should i skip the comment also?
```
#   NVMLDeviceHealthCheck: false       # GPU Health Checking using NVML
```

klueska commented on deployments/helm/nvidia-dra-driver-gpu/templates/kubeletplugin.yaml:
```diff
@@ -249,6 +249,8 @@ spec:
           value: {{ .Values.kubeletPlugin.containers.gpus.healthcheckPort | quote }}
         {{- end }}
+        - name: ADDITIONAL_XIDS_TO_IGNORE
+          value: ""
```
Drop it from the helm values, and  if someone really wants to set it they can do:
```
--set kubeletPlugin.containers.gpus.env[0].name= ADDITIONAL_XIDS_TO_IGNORE
--set kubeletPlugin.containers.gpus.env[0].value="<xid-values>"
```

klueska commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+		if err != nil {
+			_ = m.nvmllib.Shutdown()
+		}
+	}()
```
When is the shutdown called in the non-error case?

guptaNswati replied:
in `nvmlDeviceHealthMonitor.Stop()`
https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/689/files#diff-2c2a0bd3b0d6412f7f1307b6ea3cd23f451fba48563771c9cd8ad697ebf9d6c4R145

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+		supportedEvents, ret := gpu.GetSupportedEventTypes()
+		if ret != nvml.SUCCESS {
+			klog.Infof("unable to determine the supported events for %s: %v; marking it as unhealthy", parentUUID, ret)
+			m.markAllSlicesUnhealthy(giMap)
```
same questions/argument as in https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/689/files#r2577289361

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+}
+
+// markAllSlicesUnhealthy is a helper function to mark every mig device under a parent as unhealthy.
+func (m *nvmlDeviceHealthMonitor) markAllSlicesUnhealthy(giMap map[uint32]map[uint32]*AllocatableDevice) {
```
The docstring is clear, but the method name is not clear.

Let's rename to something like `markAllMigDevicesUnhealthy`.

"slice" in the context of the DRA driver in my mind refers to a resource slice.

guptaNswati replied:
done

guptaNswati commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+			// Non-blocking send to avoid deadlocks if channel is full.
+			select {
+			case m.unhealthy <- dev:
+				klog.V(6).Infof("Marked device %s as unhealthy", dev.UUID())
```
yes need to fix the log level. we recently made the default log level to 4 and so when i test this code, i have to change it to 6 to see the flow. Atleast for unhealthy events i want it be logged at the default. 

I would actually prefer to have most logs at default because in case of an unhealthy event it should be obvious as to what happened without enabling more verbose or debug

jgehrcke commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+	return m, nil
+}
+
+func (m *nvmlDeviceHealthMonitor) Start(ctx context.Context) (err error) {
```
Was this done?

guptaNswati commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+			case m.unhealthy <- dev:
+				klog.V(6).Infof("Marked device %s as unhealthy", dev.UUID())
+			default:
+				klog.Errorf("Unhealthy channel full. Dropping unhealthy notification for device %s", dev.UUID())
```
i like both ideas but need to think more about it https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/689/files/c575c63685bf6f92f71972f5c5fc737517dc77ed#r2586170468

I will add this as a follow-up To-do

guptaNswati commented on cmd/gpu-kubelet-plugin/allocatable.go:
```diff
@@ -22,11 +22,22 @@ import (
 	Gpu *GpuInfo
 	Mig *MigDeviceInfo
+	// Defined similarly as https://pkg.go.dev/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1#Healthy
+	Health HealthStatus
```
ohh, i thought `Outdated` means the suggested change is applied. 

Its done. I implemented the change and moved health to GPU and MIG structs 

```
type GpuInfo struct {
	UUID                  string `json:"uuid"`
	minor                 int
@@ -42,6 +51,7 @@ type GpuInfo struct {
	pcieBusID             string
	pcieRootAttr          *deviceattribute.DeviceAttribute
	migProfiles           []*MigProfileInfo
	Health                HealthStatus
}

type MigDeviceInfo struct {
@@ -55,6 +65,7 @@ type MigDeviceInfo struct {
	ciInfo        *nvml.ComputeInstanceInfo
	pcieBusID     string
	pcieRootAttr  *deviceattribute.DeviceAttribute
	Health        HealthStatus
}
```

guptaNswati commented on cmd/gpu-kubelet-plugin/device_health.go:
```diff
@@ -0,0 +1,342 @@
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancelContext = cancel
```
i reviewed it again. yes it redundant. i will fix it. but i think it will be useful in case of 3rd party health monitors where we will want to keep its lifecycle separate from dra driver.

guptaNswati commented:
Tested on a `GH200 1.32 cluster` by triggering an xid 43 error using null pointer access:
```
cat trigger_xid.cu 
#include <stdio.h>

// This kernel will cause a GPU page fault by writing to a null pointer.
__global__ void illegal_memory_kernel() {
    // Create a null pointer and try to write to it.
    int* bad_ptr = NULL;
    *bad_ptr = 123;
}

int main() {
    printf("Starting kernel with illegal memory access to trigger Xid...\n");

    // Launch the kernel
    illegal_memory_kernel<<<1, 1>>>();
    
    // This synchronize will now fail with an error for the bad memory access.
    cudaError_t err = cudaDeviceSynchronize();
    
    printf("Kernel finished (or was killed). Error: %s\n", cudaGetErrorString(err));
    return 0;
}
```
logs
```
I1204 20:42:19.608540       1 health.go:100] starting healthcheck service at [::]:51516
I1204 20:42:19.611278       1 device_health.go:78] creating NVML events for device health monitor
I1204 20:42:19.611301       1 device_health.go:86] registering NVML events for device health monitor
I1204 20:42:19.675749       1 device_health.go:95] started device health monitoring
I1204 20:42:19.675841       1 driver.go:228] Starting to watch for device health notifications


I1204 20:46:14.089191       1 health.go:140] Successfully invoked NodePrepareResources
I1204 20:46:19.838884       1 device_health.go:184] Processing event XID=43 event
I1204 20:46:19.838936       1 device_health.go:200] Sending unhealthy notification for device MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 due to event type:8 and event data:43
W1204 20:46:19.839016       1 driver.go:242] Received unhealthy notification for device: MIG-4d806f22-346a-5a1d-ac01-86b505cdf485
I1204 20:46:19.839029       1 device_state.go:714] Updated device: MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 health status to Unhealthy
I1204 20:46:19.839041       1 driver.go:259] Device: MIG-5ff5c47b-80e1-525d-bb61-5eef5ce0441a is healthy, added to ResoureSlice
I1204 20:46:19.839113       1 driver.go:259] Device: MIG-48e730aa-3a45-5cf9-b1f3-07d2916e9d83 is healthy, added to ResoureSlice
I1204 20:46:19.839122       1 driver.go:259] Device: MIG-d256cb8d-de32-55bf-833a-ef8a59112a84 is healthy, added to ResoureSlice
I1204 20:46:19.839129       1 driver.go:259] Device: MIG-8949ea7d-9166-53ce-9bac-6c58262addf9 is healthy, added to ResoureSlice
I1204 20:46:19.839136       1 driver.go:259] Device: MIG-24c6a0e5-fe28-53ac-9537-c4e23104f30f is healthy, added to ResoureSlice
I1204 20:46:19.839144       1 driver.go:259] Device: MIG-9b33318c-1fa4-58eb-a6f5-18273c6db866 is healthy, added to ResoureSlice
I1204 20:46:19.839152       1 driver.go:259] Device: MIG-d0c7cbd9-070b-58c6-9f87-cb8dd99a7214 is healthy, added to ResoureSlice
I1204 20:46:19.839159       1 driver.go:259] Device: GPU-9e6df7cb-64d4-5e53-2b1d-cee9e58aeb94 is healthy, added to ResoureSlice
I1204 20:46:19.839165       1 driver.go:259] Device: MIG-643fb61f-fb9a-553e-87d8-627fa535d3cb is healthy, added to ResoureSlice
I1204 20:46:19.839172       1 driver.go:259] Device: MIG-81544523-8683-58cd-8b72-5d797baf35de is healthy, added to ResoureSlice
W1204 20:46:19.839179       1 driver.go:262] Device: MIG-4d806f22-346a-5a1d-ac01-86b505cdf485 is unhealthy, will be removed from ResoureSlice
I1204 20:46:19.839182       1 driver.go:259] Device: MIG-ddcdfb36-d773-524c-9dc8-4075877523e6 is healthy, added to ResoureSlice
I1204 20:46:19.839192       1 driver.go:259] Device: GPU-a4f34abc-7715-3560-dcea-7238b9611a45 is healthy, added to ResoureSlice
I1204 20:46:19.839198       1 driver.go:259] Device: MIG-4e9fed73-e2a4-512f-ad8e-e7bb75aa920e is healthy, added to ResoureSlice
I1204 20:46:19.839206       1 driver.go:259] Device: MIG-c7cbc24b-a93f-5ff3-b388-deba1f3ac17e is healthy, added to ResoureSlice
I1204 20:46:19.839214       1 driver.go:259] Device: MIG-59e178f7-0c5e-52e3-902c-bf2f31c2275f is healthy, added to ResoureSlice
I1204 20:46:19.839221       1 driver.go:266] Rebulishing resourceslice with healthy devices
I1204 20:46:19.839267       1 driver.go:287] Successfully republished resources without unhealthy device
```
quick check on unhealthy device entry vs healthy:
```
$ kubectl get resourceslice sc-starwars-mab9-b00-gpu.nvidia.com-6h999 -o yaml | grep MIG-4d806f22-346a-5a1d-ac01-86b505cdf485
$  kubectl get resourceslice sc-starwars-mab9-b00-gpu.nvidia.com-6h999 -o yaml | grep MIG-5ff5c47b-80e1-525d-bb61-5eef5ce0441a
          string: MIG-5ff5c47b-80e1-525d-bb61-5eef5ce0441a**
```

guptaNswati commented:
> 🚀 What a journey :-). Thanks everyone for all the help and input and ideas that we threw together. And thank you Swati for bearing with us.
> 
> I just saw green CI. Let's now merge this, and then let's iterate after merge.

Oh i agree. Thank you @jgehrcke and thank you everyone. Alot of learnings here and follow-up improvements. Lets prioritize these early next year. With all the discussions, I realized how imp is this feature and needs to be done more thoroughly.

elezar commented:
Thanks @guptaNswati!

Just a note on the XID 43 that we're triggering with the example. According to the [documentation](https://docs.nvidia.com/deploy/xid-errors/analyzing-xid-catalog.html#step-4-determine-resolution-steps) this XID is an `ROBUST_CHANNEL_RESETCHANNEL_VERIF_ERROR` and should be ignored.

> "This event is logged when a user application hits a software induced fault and must terminate. The GPU remains in a healthy state.

> In most cases, this is not indicative of a driver bug but rather a user application error."

It is also in the list of XIDs that we explicitly [skip](https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/64b6718fd2c2d37754eb71e2119e8a0fc86b6782/cmd/gpu-kubelet-plugin/device_health.go#L327). Did you run the test with some other configuration to trigger this?
